eclipsesource/theia
# Support tree view drag and drop.

Signed-off-by: Thomas Mäder <t.s.maeder@gmail.com>



#### What it does
Adds support for contributing drag and drop controller in the tree view plugin API. The code tries to mimick the behaviour described in https://code.visualstudio.com/api/references/vscode-api#TreeDragAndDropController&lt;T&gt; as well as possible. I have also refactored some of the "drag into editor area" code to work with the new code (and fixed dragging external files into electron on windows as a "drive-by")

Contributed on behalf of ST Microelectronics

Fixes #11778, Fixes #11776

#### How to test
I have adapted the VS Code tree view sample to exercise the code added in the PR. The extension & source is attached to this PR.
The extension contributes a view called "Test View Drag and Drop", which you need to open.

The example implement multiple features:
1. Drag and drop within the tree
The dropped element should be moved under the element where it was dropped.
Note that there seems to be a bug in our tree-view refresh code where the UI goes into an endless loop when you reparent the elements of the tree: It seems the parent field of the tree items is not properly updated when an item is reparented. I am pretty convinced this is not related to the changes in this PR.
2. Drag an element to the editor area:
This should open a *.js file from the extension itself: note that the cursor will be placed at line 10, column 6.
3. Exercise the current drag and drop to the editor area. Drag from the file navigator and from the OS file explorer
4. Exercise dropping files to the example view: the extension will show the first couple of bytes of the file content.

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
